yt/yql/agent/type_builder.cpp

```diff
@@ -0,0 +1,334 @@
+    Push(SimpleLogicalType(ESimpleLogicalValueType::Date32));
+}
+
+void TTypeBuilder::OnDatetime64()
+{
+    Push(SimpleLogicalType(ESimpleLogicalValueType::Datetime64));
+}
+
+void TTypeBuilder::OnTimestamp64()
+{
+    Push(SimpleLogicalType(ESimpleLogicalValueType::Timestamp64));
+}
+
+void TTypeBuilder::OnTzDate32()
+{
+    Push(SimpleLogicalType(ESimpleLogicalValueType::String));
+}
+
+void TTypeBuilder::OnTzDatetime64()
+{
+    Push(SimpleLogicalType(ESimpleLogicalValueType::String));
+}
+
+void TTypeBuilder::OnTzTimestamp64()
+{
+    Push(SimpleLogicalType(ESimpleLogicalValueType::String));
+}
+
+void TTypeBuilder::OnInterval64()
+{
+    Push(SimpleLogicalType(ESimpleLogicalValueType::Interval64));
+}
+
+void TTypeBuilder::OnDecimal(ui32 precision, ui32 scale)
+{
+    Push(DecimalLogicalType(precision, scale));
+}
+
+void TTypeBuilder::OnBeginOptional()
+{
+    ItemsStack.emplace(TItemType());
+}
+
+void TTypeBuilder::OnEndOptional()
+{
+    Push(OptionalLogicalType(Pop()));
+}
+void TTypeBuilder::OnBeginList()
+{
+    ItemsStack.emplace(TItemType());
+}
+
+void TTypeBuilder::OnEndList()
+{
+    Push(ListLogicalType(Pop()));
+}
+
+void TTypeBuilder::OnBeginTuple()
+{
+    ItemsStack.emplace(TElements());
+}
+
+void TTypeBuilder::OnTupleItem()
+{
+}
+
+void TTypeBuilder::OnEndTuple()
+{
+    Push(TupleLogicalType(Pop<TElements>()));
+}
+
+void TTypeBuilder::OnBeginStruct()
+{
+    ItemsStack.emplace(TMembers());
+}
+void TTypeBuilder::OnStructItem(TStringBuf member)
+{
+    MemberNames.emplace(member);
+}
+
+void TTypeBuilder::OnEndStruct()
+{
+    Push(StructLogicalType(Pop<TMembers>()));
+}
+
+void TTypeBuilder::OnBeginDict()
+{
+    ItemsStack.emplace(TKeyAndPayload());
+}
+
+void TTypeBuilder::OnDictKey()
+{
+    std::get<TKeyAndPayload>(ItemsStack.top()).Switch = true;
+}
+
+void TTypeBuilder::OnDictPayload()
+{
+    std::get<TKeyAndPayload>(ItemsStack.top()).Switch = false;
+}
+
+void TTypeBuilder::OnEndDict()
+{
+    auto items = Pop<TKeyAndPayload>();
+    Push(DictLogicalType(std::move(items.Key), std::move(items.Payload)));
+}
+
+void TTypeBuilder::OnBeginVariant()
+{
+    ItemsStack.emplace(TItemType());
+}
+
+void TTypeBuilder::OnEndVariant()
+{
+    const auto internal = Pop();
+    switch (internal->GetMetatype()) {
+        case ELogicalMetatype::Struct:
+            Push(VariantStructLogicalType(internal->GetFields()));
+            break;
+        case ELogicalMetatype::Tuple:
+            Push(VariantTupleLogicalType(internal->GetElements()));
+            break;
+        default:
+            THROW_ERROR_EXCEPTION("Invalid variant type.");
+    }
+}
+
+void TTypeBuilder::OnBeginTagged(TStringBuf tag)
+{
+    ItemsStack.emplace(TTagAndType(tag, nullptr));
+}
+
+void TTypeBuilder::OnEndTagged()
+{
+    auto [tag, type] = Pop<TTagAndType>();
+    Push(TaggedLogicalType(tag, std::move(type)));
+}
+
+void TTypeBuilder::OnPg(TStringBuf name, TStringBuf category)
+{
+    THROW_ERROR_EXCEPTION("%s not implemented.", __func__);
```

Comment:
Why do we have an error here?

Reply:
I planned to add the next PR if everything else is ok with this one.